### PR TITLE
damlc: Don't rely on worker for zero argument polymorphic type constructors

### PR DESCRIPTION
**Pull request by sofiafaro-da**
Fixes https://github.com/digital-asset/daml/issues/12166 which was caused by using a constructor worker (`$WBaz`) in a case where GHC doesn't generate them. This only caused a problem with partially applied zero argument constructors, so, it only affected zero argument constructors of polymorphic types in situations where the type argument is eta-reduced, which is probably why it flew under the radar.

The code in the bug report is added as a regression test.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs [hunk 1]
```diff
@@ -1729,23 +1748,26 @@ convertDataCon env m con args
...
-    where
-
-        qual :: (T.Text -> n) -> T.Text -> ConvertM (Qualified n)
-        qual f t
-            | Just xs <- T.stripPrefix "(," t
-            , T.dropWhile (== ',') xs == ")" = qDA_Types env $ f $ "Tuple" <> T.pack (show $ T.length xs + 1)
-            | IgnoreWorkerPrefix t' <- t = qualify env m $ f t'
+    qual :: (T.Text -> n) -> T.Text -> ConvertM (Qualified n)
+    qual f t
+        | Just xs <- T.stripPrefix "(," t
```
Comment (cocreature) [inline compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs:1753]: this code freaks me out everytime I see it after I’ve forgotten it exists.

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs [hunk 2]
```diff
@@ -1698,14 +1698,33 @@ convertDataCon env m con args
     | Just (tyArgs, tmArgs) <- splitConArgs_maybe con args = do
         tyArgs <- mapM (convertType env) tyArgs
         tmArgs <- mapM (convertExpr env) tmArgs
+        (, []) <$> fullyApplied env tyArgs tmArgs
+
+    -- Partially applied, but the constructor only takes type args.
+    -- In this situation there is no worker function, so we inline
+    -- the constructor by introducing some type lambdas.
+    | let (conTypes, conTheta, conArgs, _) = dataConSig con
+    , length conTheta + length conArgs == 0 -- no value args
```
Comment (cocreature) [inline compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs:1707]:
```suggestion
    , null conTheta && null conArgs -- no value args
```
maybe a bit clearer? (and potentially slightly more efficient)

**Review by cocreature — APPROVED** [review]
Great work, thanks a lot!

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs (hunk 2 above)
Comment (sofiafaro-da) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs:1707]: ah, nice, yes :)

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs (hunk 1 above)
Comment (sofiafaro-da) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs:1753]: 🙃

Comment (akrmn) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs:1753]: 😮

**Pull request merged**